microprofile/WFLY-10711_health_check_extension.adoc

```diff
@@ -0,0 +1,99 @@
+* mailto:tremes@redhat.com[Tomas Remes] - OpenShift
+
+=== Affected Projects or Components
+
+The implementation of Eclipse MicroProfile Health is provided by https://github.com/smallrye/smallrye-health[smallrye-health].
+
+=== Other Interested Projects
+
+* A https://github.com/jmesnil/wildfly-microprofile-health[prototype of a WildFly extension] for MicroProfile Health was developed during the development of MicroProfile Health 1.0.
+
+== Requirements
+
+* Support Eclipse MicroProfile Health 1.0 for applications deployed in WildFly.
+* Provide a management runtime operation to check the health of the application server and its deployed applications. The result of the runtime operation
+  will be the DRM representation of the https://github.com/eclipse/microprofile-health/blob/master/spec/src/main/asciidoc/protocol-wireformat.adoc#appendix-b-json-payload-specification[MicroProfile Health JSON payload]
+* Provide an optional HTTP endpoint for the https://github.com/eclipse/microprofile-health/blob/master/spec/src/main/asciidoc/protocol-wireformat.adoc[health wireformat].
+  If the HTTP endpoint is enabled, it requires the presence of the WildFly https://github.com/wildfly/wildfly-capabilities/blob/master/org/wildfly/management/http-interface/capability.adoc[HTTP management interface capability]
+* The HTTP endpoint is global to the server: it must aggregate health check probes from *all deployments*.
+
+== Nice-to-Have Requirements
+
+* Support drilldown of the HTTP endpoint (e.g. to be able to query the health of individual deployment, e.g. `/health/deploymentA`)
+* Provide a WildFly management API so that other extensions can define health check probes that taken into account
+  by the HTTP endpoint (this is out of scope of this proposal but the design of the new extension must not preclude to
+  provide such an API at a later point).
+
+== Non-Requirements
+
+* MicroProfile Health 1.0 only provides a "liveness" probe (as defined by https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-liveness-command[Kubernetes]).
+__A Readiness probe is not covered by this proposal__.
+
+== Implementation Plan
+
+* Add dependencies to `io.smallrye:smallrye-health` and `org.eclipse.microprofile.health:microprofile-health-api:1.0` artifacts
+* Add a new `microprofile-health-smallrye` extension to provide Health support for deployments.
+* Provide a Undertow handler that exposes the optional health HTTP endpoint `/health` on WildFly `management-http` socket binding (port `9990` of the `management` network interface).
+
+=== Subsystem description
+
+The subsystem will be named `microprofile-health-smallrye`.
+It will define 1 runtime operation:
+
+* `check` whose result is the DRM representation of the Health Check JSON payload. If the global outcome is `UP`, the result's `outcome` will be `success` and the `result` will be the
+DMR representation of the HTTP JSON payload. Otherwise, the operation will fail.
```

Comment:
Do we need to have such operation ? 

`curl http://SERVER:9990/health` or `firefox http://SERVER:9990/health` seems sufficient

Reply:
An operation like this might be useful in the openshift integration, yes.  All of our current liveness probes are based on management requests to the http interface, with the different requests done in a composite/batch. Adding another operation to the batch is an easy way to incorporate this with the overall liveness probe.

Reply:
@bstansberry Having the management operation fails when the health check outcome is DOWN is consistent with the HTTP wireformat (200 for `UP` and 500/503 for `DOWN`) and [Kubernetes Liveness command](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#before-you-begin):

> If the command succeeds, it returns 0, and the kubelet considers the Container to be alive and healthy. If the command returns a non-zero value, the kubelet kills the Container and restarts it.

Making the `check-health` operation fails ensure that we do not report false positive to Kubernetes (as it is only interested by the return result of the command).

Data returned by the operation are informational and will help users analyse ongoing liveness issues but from operational point of view, they do not matter for OpenShift.
The check may fail because the subsystem is not installed but I'd argue this should also be part of the healthiness of the application server.

Reply:
I'm not convinced. :)

For sure I disagree that the presence of this subsystem is required for the server to be considered healthy.  Maybe some day if we start integrating other subsystem checks into the response we could say that, but right now it's only app checks and if the dev hasn't written any, then this subsystem is adding nothing.  We already have liveness and readiness checks; this is just adding another and shouldn't break what's already there. 

We'd ship it by default in our OS image configs, so if it's not present in an image version that supports it that means the user chose not to include it. Maybe that was a mistake on their part, as their app includes checks they've disabled, but I don't think we should fail liveness on the server because maybe they made the wrong choice. Maybe they knew what they were doing.

As a practical matter, having to use forked liveness probe code for images that support this feature vs those that do not is a pretty big hassle. It's much better to have the same probe code be able to discriminate the "not applicable" case from the DOWN case.

The outcome=success vs failure part I'm not convinced but I feel much less strongly about it. If we write our probes to use this DMR op I think we'd send a composite, a simple read-resource step and a check step. A failure in the r-r step would indicate "not applicable", so then the outcome of the check step is only meaningful if the r-r succeeds.

I don't think the HTTP wireformat requirement nor the liveness command return code control this. Particularly not the latter as there has to be non-HTTP-oriented logic between a DMR response and the return code of any command process that made the DMR request.

Still, I don't feel that strongly about it.

BTW https://issues.jboss.org/browse/WFCORE-185 would result in a 404 coming from the "REST" (i.e. HTTP, non-Upgrade) management interface for a read of a non-existent resource. This isn't directly relevant to this discussion though, it's just semi-related.

Related: if the /health context does not exist on the HTTP interface then I think the response should be a 404.  I believe that is what would happen and I think it's right.